File: abstractions/java/lib/src/main/java/com/microsoft/kiota/serialization/SerializationWriterFactoryRegistry.java
package com.microsoft.kiota.serialization;

import java.util.HashMap;
import java.util.Objects;
import java.util.regex.Pattern;

import javax.annotation.Nonnull;
/** This factory holds a list of all the registered factories for the various types of nodes. */
public class SerializationWriterFactoryRegistry implements SerializationWriterFactory {
    /** Default singleton instance of the registry to be used when registring new factories that should be available by default. */
    public final static SerializationWriterFactoryRegistry defaultInstance = new SerializationWriterFactoryRegistry();
    /** List of factories that are registered by content type. */
    public HashMap<String, SerializationWriterFactory> contentTypeAssociatedFactories = new HashMap<>();
    public String getValidContentType() {
        throw new UnsupportedOperationException("The registry supports multiple content types. Get the registered factory instead.");
    }
    private static Pattern contentTypeVendorCleanupPattern = Pattern.compile("[^/]+\\+", Pattern.CASE_INSENSITIVE);
    @Override
    @Nonnull
    public SerializationWriter getSerializationWriter(@Nonnull final String contentType) {
        Objects.requireNonNull(contentType, "parameter contentType cannot be null");
        if(contentType.isEmpty()) {
            throw new NullPointerException("contentType cannot be empty");
        }
        final String vendorSpecificContentType = contentType.split(";")[0];
        if(contentTypeAssociatedFactories.containsKey(vendorSpecificContentType)) {
            return contentTypeAssociatedFactories.get(vendorSpecificContentType).getSerializationWriter(vendorSpecificContentType);
        }
        final String cleanedContentType = contentTypeVendorCleanupPattern.matcher(vendorSpecificContentType).replaceAll("");
        if(contentTypeAssociatedFactories.containsKey(cleanedContentType)) {
            return contentTypeAssociatedFactories.get(cleanedContentType).getSerializationWriter(cleanedContentType);
        }
        throw new RuntimeException("Content type " + contentType + " does not have a factory to be serialized");
    }
    
}

File: abstractions/python/kiota/abstractions/serialization/serialization_writer_factory_registry.py
import re
from typing import Dict

from .serialization_writer import SerializationWriter
from .serialization_writer_factory import SerializationWriterFactory


class SerializationWriterFactoryRegistry(SerializationWriterFactory):
    """This factory holds a list of all the registered factories for the various types of nodes.
    """
    # List of factories that are registered by content type.
    CONTENT_TYPE_ASSOCIATED_FACTORIES: Dict[str, SerializationWriterFactory] = {}

    __instance = None

    def __new__(cls, *args, **kwargs):
        """Default singleton instance of the registry to be used when registring new
        factories that should be available by default.

        Returns:
            [SerializationWriterFactoryRegistry]: Default singleton instance of the class
        """
        if not SerializationWriterFactoryRegistry.__instance:
            SerializationWriterFactoryRegistry.__instance = object.__new__(cls)
        return SerializationWriterFactoryRegistry.__instance

    def get_valid_content_type(self) -> str:
        raise Exception(
            "The registry supports multiple content types. Get the registered factory instead"
        )

    def get_serialization_writer(self, content_type: str) -> SerializationWriter:
        if not content_type:
            raise Exception("Content type cannot be null")

        vendor_specific_content_type = content_type.split(';')[0]
        factory = self.CONTENT_TYPE_ASSOCIATED_FACTORIES.get(vendor_specific_content_type)
        if factory:
            return factory.get_serialization_writer(vendor_specific_content_type)
        cleaned_content_type = re.sub(r'[^/]+\+', '', vendor_specific_content_type)

        factory = self.CONTENT_TYPE_ASSOCIATED_FACTORIES.get(cleaned_content_type)
        if factory:
            return factory.get_serialization_writer(cleaned_content_type)
        raise Exception(
            f"Content type {cleaned_content_type} does not have a factory registered"
            "to be serialized"
        )
